ext/crypto/00_crypto.js

```diff
@@ -141,6 +146,215 @@
     return StringPrototypeReplace(base64String, /=/g, "");
   }
 
+  function isArrayOrArrayBufferView(obj) {
+    return ArrayIsArray(obj) || ArrayBufferIsView(obj);
+  }
+
+  function concatBytes(...bytes) {
+    return Uint8Array.from(
+      bytes.reduce(
+        (a, val) =>
+          (isArrayOrArrayBufferView(val)) ? [...a, ...val] : [...a, val],
+        [],
+      ),
+    );
+  }
+
+  function buildTLV(tag, value) {
+    if (isArrayOrArrayBufferView(value)) {
+      value = concatBytes(...value);
+    }
+    const len = value.length;
+    const lenBytes = [];
+
+    if (len <= 127) {
+      lenBytes.push(len);
+    } else if (len < 256) {
+      lenBytes.push(0x81);
+      lenBytes.push(len);
+    } else if (len < 65536) {
+      lenBytes.push(0x82);
+      lenBytes.push(len >> 8);
+      lenBytes.push(len & 0xff);
+    }
+
+    return concatBytes(tag, lenBytes, value);
+  }
+
+  function buildIntegerTLV(val) {
+    return buildTLV(0x02, (val[0] & 0x80) ? [0x00, ...val] : val);
+  }
+
+  function buildRsaPkcs(key, type) {
+    const algo = buildTLV(0x30, [
+      buildTLV(0x06, [0x2A, 0x86, 0x48, 0x86, 0xF7, 0x0D, 0x01, 0x01, 0x01]),
+      buildTLV(0x05, []),
+    ]);
+
+    const modExp = concatBytes(buildIntegerTLV(key.n), buildIntegerTLV(key.e));
+
+    if (type == "pkcs8") {
+      let keyData = concatBytes(
+        buildIntegerTLV([0x00]),
+        modExp,
+        buildIntegerTLV(key.d),
+      );
+
+      if (key.p) {
+        keyData = concatBytes(
+          keyData,
+          buildIntegerTLV(key.p),
+          buildIntegerTLV(key.q),
+          buildIntegerTLV(key.dp),
+          buildIntegerTLV(key.dq),
+          buildIntegerTLV(key.qi),
+        );
+      }
+
+      return buildTLV(0x30, [
+        buildTLV(0x02, [0x00]), // version
+        algo,
+        buildTLV(0x04, buildTLV(0x30, keyData)),
+      ]);
+    } else {
+      return buildTLV(0x30, [
+        algo,
+        buildTLV(0x03, concatBytes(0x00, buildTLV(0x30, modExp))),
+      ]);
+    }
+  }
+
+  function parseTLV(bytes) {
```

Comment:
The `rsa` crate already comes with a ASN.1 DER decoder/encoder. It's better to use those instead of rolling a JavaScript implementation.